aiohttp/web_urldispatcher.py

```diff
@@ -1028,44 +1099,65 @@ def register_resource(self, resource: AbstractResource) -> None:
-            parts = self.NAME_SPLIT_RE.split(name)
-            for part in parts:
-                if keyword.iskeyword(part):
-                    raise ValueError(
-                        f"Incorrect route name {name!r}, "
-                        "python keywords cannot be used "
-                        "for route name"
-                    )
-                if not part.isidentifier():
-                    raise ValueError(
-                        "Incorrect route name {!r}, "
-                        "the name should be a sequence of "
-                        "python identifiers separated "
-                        "by dash, dot or column".format(name)
-                    )
-            if name in self._named_resources:
+        if not name:
+            raise ValueError("Cannot register a named resource with empty name.")
+
+        parts = self.NAME_SPLIT_RE.split(name)
+        for part in parts:
+            if keyword.iskeyword(part):
                 raise ValueError(
-                    "Duplicate {!r}, "
-                    "already handled by {!r}".format(name, self._named_resources[name])
+                    f"Incorrect route name {name!r}, "
+                    "python keywords cannot be used "
+                    "for route name"
                 )
-            self._named_resources[name] = resource
-        self._resources.append(resource)
+            if not part.isidentifier():
+                raise ValueError(
+                    "Incorrect route name {!r}, "
+                    "the name should be a sequence of "
+                    "python identifiers separated "
+                    "by dash, dot or column".format(name)
+                )
+        if name in self._named_resources:
+            raise ValueError(
+                "Duplicate {!r}, "
+                "already handled by {!r}".format(name, self._named_resources[name])
+            )
+        self._named_resources[name] = resource
 
     def add_resource(self, path: str, *, name: Optional[str] = None) -> Resource:
         if path and not path.startswith("/"):
             raise ValueError("path should be started with / or be empty")
+        # Reuse named resource
+        if name:
+            resource = self._named_resources.get(name)
+            if resource is None:
+                pass
+            elif resource.raw_match(path):
+                return cast(Resource, resource)
+            else:
+                raise ValueError(f"Duplicate {name!r}, already handled by {resource!r}")
+        if not ("{" in path or "}" in path or ROUTE_RE.search(path)):
+            if self._resources and isinstance(self._resources[-1], GroupPlainResource):
+                grp = self._resources[-1]
+            else:
+                grp = GroupPlainResource()
+                self.register_resource(grp)
+            resource = grp.add_resource(_requote_path(path), name=name)
+            if name:
+                self._register_named_resource(resource)
```

Comment:
I think there's an edge case here where the `frozen` check doesn't happen. I'd like to see a test for that, I think something like:
```
add_resource("/foo", "foo")
freeze()
add_resource("/bar", "bar")
```

I suspect a simple solution to then get the test passing would be to ensure the group is frozen. It looks like `grp.add_resource()` currently will allow adding new resources after `grp.freeze()`, which seems incorrect. It freezes the resources, but not itself.

Reply:
Covered by test_frozen_plain_group